### PR TITLE
cleanly separate flags and options in zio package

### LINKED ISSUE #1250: refactor zio reader/writer options and flags and improve CLI tools re-use of flags

### PR DESCRIPTION
**Pull request by mccanne**
This commit eliminates significant technical debt for how we
manage command flags and configuration options for the zio
readers and writers.  There were multiple patterns fighting
agaist each other.

This commit cleanly separates command flags from zio options
and refactors the CLI tools to re-use many common flags.
For example, you can now set the sortmem parameter and run
profiling from most all of the tools.

This also cleans up the way json type mappings are read and
configured into the reader options.

In the new approach, each zio reader/writer is responsible for
creating an options struct and package zio/options glues them
together.  The only package under zio that knows about command
flags is zio/flags, and its job is to relate flags to configuration
options.

We also deleted the "types" output format rather than make it work
here since it was hokey (my bad) and I think a better approach
is to make proc that helps collect up all the types.

Whether and when we switch over to cobra (or something else), these
changes are orthogonal and still very helpful.

Fixes #1250

### CODE REVIEW

**Comment by philrz** [comment]
@mccanne: As long as you're making common flags work across the tools, is now the time to ask about getting a `-version` flag in all of them? 😄

**Comment by mccanne** [comment]
> @mccanne: As long as you're making common flags work across the tools, is now the time to ask about getting a `-version` flag in all of them? 😄

Piece of cake!

**Comment by mccanne** [comment]
@philrz I just pushed commit to make -version common so it's now in all the tools.

File: cmd/cli/cli.go [hunk 1]
```diff
@@ -0,0 +1,155 @@
+package cli
```
Comment (nwt) [inline cmd/cli/cli.go:1]: A non-main package in cmd/ is a bit of a surprise. I think this should go elsewhere.

File: zio/flags/flags.go [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+	"regexp"
+
+	"github.com/brimsec/zq/zio/ndjsonio"
+	"github.com/brimsec/zq/zio/options"
+	"github.com/brimsec/zq/zio/zngio"
+)
+
+// Reader has the union of all the flags accepted by the different
+// Reader implementations.
+type Reader struct {
```
Comment (nwt) [inline zio/flags/flags.go:17]: Nit: As in zio/options, calling these types Reader and Writer feels off. I'd call them ReaderFlags and WriterFlags.

File: cmd/cli/cli.go [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+func IsTerminal(f *os.File) bool {
+	return terminal.IsTerminal(int(f.Fd()))
+}
+
+func runCPUProfile(path string) {
+	f, err := os.Create(path)
+	if err != nil {
+		log.Fatal(err)
+	}
+	pprof.StartCPUProfile(f)
```
Comment (nwt) [inline cmd/cli/cli.go:144]: We never close f, but I think that's fine.

**Review by nwt — APPROVED** [review]
My comments are all nits, so I think this is fine to merge.

File: zio/options/options.go [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+	JSON   ndjsonio.ReaderOpts
+	AwsCfg *aws.Config
+}
+
+type Writer struct {
+	Format string
+	UTF8   bool
+	Text   textio.WriterOpts
+	Zng    zngio.WriterOpts
+}
```
[the remark this answers is not in the document]
Comment (mccanne) [reply zio/options/options.go:22]: They can't go in zio because of circular dependency.

File: cmd/cli/cli.go (hunk 1 above)
Comment (mccanne) [reply cmd/cli/cli.go:1]: I'll put it in brimsec/zq/cli

File: zio/options/options.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (mccanne) [reply zio/options/options.go:22]: I am not happy with alternative naming schemes I'm coming up with right now.  I'll merge it this way and make an issue to revisit zio/flags and zio/options.  I guess I don't like zio/zioflags, but maybe I would get used to it.  We can bikeshed a bit and fix it in a later PR.

File: zio/flags/flags.go [hunk 5]
```diff
@@ -0,0 +1,85 @@
+package flags
```
[the remark this answers is not in the document]
Comment (mccanne) [reply zio/flags/flags.go:1]: Agreed.  See issue #1253

File: cmd/cli/cli.go (hunk 3 above)
Comment (mccanne) [reply cmd/cli/cli.go:144]: Good point.  Probably not a problem, but I'll fix.  I think it was that way before.

File: zio/flags/flags.go (hunk 2 above)
Comment (mccanne) [reply zio/flags/flags.go:17]: Agreed.  See issue #1253

**Pull request merged**